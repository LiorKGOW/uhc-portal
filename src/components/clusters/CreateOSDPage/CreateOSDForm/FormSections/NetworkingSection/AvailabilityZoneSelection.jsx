// a redux-form Field-compatible compontent for selecting an availability zone

import React from 'react';
import PropTypes from 'prop-types';
import {
  Select, SelectOption, FormGroup,
} from '@patternfly/react-core';

const PLACEHOLDER_VALUE = 'Select availability zone';

class AvailabilityZoneSelection extends React.Component {
  state = { isOpen: false };

  onToggle = (isOpen) => {
    this.setState({ isOpen });
  }

  onSelect = (_, selection) => {
    const { input } = this.props;
    this.setState({ isOpen: false });
    if (selection === PLACEHOLDER_VALUE) {
      input.onChange('');
    } else {
      input.onChange(selection);
    }
  }

  render() {
    const { isOpen } = this.state;
    const {
      input, isDisabled, region, label, meta: { error, touched },
    } = this.props;
    return (
      <FormGroup
        {...input}
        label={label}
        className="ocm-c-create-osd-az-select"
        validated={error ? 'error' : undefined}
        helperTextInvalid={touched && error}
      >
        <Select
          isOpen={isOpen}
          selections={input.value || PLACEHOLDER_VALUE}
          onToggle={this.onToggle}
          onSelect={this.onSelect}
          isDisabled={isDisabled}
        >
          <SelectOption key={0} value={PLACEHOLDER_VALUE} isPlaceholder />
          {['a', 'b', 'c', 'd', 'e', 'f'].map((letter) => {
            // AWS availability zones are comprised from the region name
            // followed by a single letter. For more information please see:
            // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-availability-zones
            const avalabilityZone = region + letter;
            return <SelectOption key={letter} value={avalabilityZone}>{`${avalabilityZone}`}</SelectOption>;
          })}
        </Select>
      </FormGroup>
    );
  }
}
AvailabilityZoneSelection.propTypes = {
  isDisabled: PropTypes.bool,
  label: PropTypes.string,
  region: PropTypes.string,
  input: PropTypes.shape({
    value: PropTypes.string,
    onChange: PropTypes.func,
  }),
  meta: PropTypes.shape({
    touched: PropTypes.bool,
    error: PropTypes.string,
  }),
};

export default AvailabilityZoneSelection;
