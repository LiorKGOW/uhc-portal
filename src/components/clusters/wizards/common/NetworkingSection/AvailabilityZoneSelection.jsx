// a redux-form Field-compatible component for selecting an availability zone

import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup } from '@patternfly/react-core';
import {
  Select as SelectDeprecated,
  SelectOption as SelectOptionDeprecated,
} from '@patternfly/react-core/deprecated';
import { FormGroupHelperText } from '~/components/common/FormGroupHelperText';

// AWS availability zones are comprised from the region name
// followed by a single letter. For more information please see:
// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-availability-zones
const azLetters = ['a', 'b', 'c', 'd', 'e', 'f'];

class AvailabilityZoneSelection extends React.Component {
  state = { isOpen: false };

  onToggle = (isOpen) => {
    this.setState({ isOpen });
  };

  onSelect = (_, selection) => {
    const { input } = this.props;
    this.setState({ isOpen: false });
    input.onChange(selection);
  };

  render() {
    const { isOpen } = this.state;
    const {
      input,
      isDisabled,
      enabledAvailabilityZones,
      region,
      label,
      meta: { error, touched },
    } = this.props;

    const availabilityZones = azLetters
      .map((letter) => ({
        zoneId: `${region}${letter}`,
        isDisabled:
          enabledAvailabilityZones && !enabledAvailabilityZones.includes(`${region}${letter}`),
      }))
      .sort((azA, azB) => {
        if (azA.isDisabled && !azB.isDisabled) {
          return 1;
        }
        if (azB.isDisabled && !azA.isDisabled) {
          return -1;
        }
        return azA.zoneId.localeCompare(azB.zoneId);
      });

    return (
      <FormGroup
        fieldId={input.name}
        label={label}
        className="ocm-c-create-osd-az-select"
        isRequired
      >
        <SelectDeprecated
          isOpen={isOpen}
          selections={input.value}
          onToggle={(_event, isOpen) => this.onToggle(isOpen)}
          onSelect={this.onSelect}
          isDisabled={isDisabled}
          placeholderText="Select availability zone"
          aria-label={label}
        >
          {availabilityZones.map(({ zoneId, isDisabled }) => (
            <SelectOptionDeprecated
              key={zoneId}
              value={zoneId}
              isDisabled={isDisabled}
              description={isDisabled ? 'This zone does not have all required subnets' : undefined}
            >
              {zoneId}
            </SelectOptionDeprecated>
          ))}
        </SelectDeprecated>

        <FormGroupHelperText touched={touched} error={error} />
      </FormGroup>
    );
  }
}
AvailabilityZoneSelection.propTypes = {
  isDisabled: PropTypes.bool,
  label: PropTypes.string,
  region: PropTypes.string,
  enabledAvailabilityZones: PropTypes.arrayOf(PropTypes.string),
  input: PropTypes.shape({
    name: PropTypes.string,
    value: PropTypes.string,
    onChange: PropTypes.func,
  }),
  meta: PropTypes.shape({
    touched: PropTypes.bool,
    error: PropTypes.string,
  }),
};

export default AvailabilityZoneSelection;
