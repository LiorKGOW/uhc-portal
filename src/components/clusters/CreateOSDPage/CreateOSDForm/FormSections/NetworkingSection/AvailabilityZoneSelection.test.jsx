import React from 'react';
import { shallow } from 'enzyme';
import AvailabilityZoneSelection from './AvailabilityZoneSelection';

describe('<AvailabilityZoneSelection />', () => {
  const onChange = jest.fn();
  const wrapper = shallow(
    <AvailabilityZoneSelection
      label="some label"
      region="fake-region"
      input={{
        value: '',
        onChange,
      }}
      meta={{}}
    />,
  );

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should show error on error', () => {
    wrapper.setProps({ meta: { touched: true, error: 'some error message' } });
    const formGroupProps = wrapper.find('FormGroup').props();
    expect(formGroupProps.helperTextInvalid).toEqual('some error message');
    expect(formGroupProps.validated).toEqual('error');
  });
});
