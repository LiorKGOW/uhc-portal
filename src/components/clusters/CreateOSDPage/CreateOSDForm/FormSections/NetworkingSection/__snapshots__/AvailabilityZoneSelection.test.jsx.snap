// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvailabilityZoneSelection /> should render 1`] = `
<FormGroup
  className="ocm-c-create-osd-az-select"
  isRequired={true}
  label="some label"
  onChange={[MockFunction]}
  value=""
>
  <Select
    aria-describedby=""
    aria-invalid={false}
    aria-label=""
    aria-labelledby=""
    chipGroupComponent={null}
    className=""
    clearSelectionsAriaLabel="Clear all"
    createText="Create"
    customBadgeText={null}
    customContent={null}
    direction="down"
    favorites={Array []}
    favoritesLabel="Favorites"
    hasInlineFilter={false}
    hasPlaceholderStyle={false}
    inlineFilterPlaceholderText={null}
    inputAutoComplete="off"
    inputIdPrefix=""
    isCreatable={false}
    isCreateOptionOnTop={false}
    isCreateSelectOptionObject={false}
    isDisabled={false}
    isFlipEnabled={true}
    isGrouped={false}
    isInputFilterPersisted={false}
    isInputValuePersisted={false}
    isOpen={false}
    isPlain={false}
    menuAppendTo="inline"
    noResultsFoundText="No results found"
    onClear={[Function]}
    onCreateOption={[Function]}
    onFilter={null}
    onSelect={[Function]}
    onToggle={[Function]}
    onTypeaheadInputChanged={null}
    ouiaSafe={true}
    placeholderText=""
    position="left"
    removeSelectionAriaLabel="Remove"
    selections="Select availability zone"
    shouldResetOnSelect={true}
    toggleAriaLabel="Options menu"
    toggleIcon={null}
    toggleId={null}
    toggleIndicator={null}
    typeAheadAriaDescribedby=""
    typeAheadAriaLabel=""
    validated="default"
    variant="single"
    width=""
    zIndex={9999}
  >
    <SelectOption
      className=""
      component="button"
      index={0}
      inputId=""
      isChecked={false}
      isDisabled={false}
      isFavorite={null}
      isLastOptionBeforeFooter={[Function]}
      isLoad={false}
      isLoading={false}
      isNoResultsOption={false}
      isPlaceholder={true}
      isSelected={false}
      key="0"
      keyHandler={[Function]}
      onClick={[Function]}
      sendRef={[Function]}
      setViewMoreNextIndex={[Function]}
      value="Select availability zone"
    />
    <SelectOption
      className=""
      component="button"
      index={0}
      inputId=""
      isChecked={false}
      isDisabled={false}
      isFavorite={null}
      isLastOptionBeforeFooter={[Function]}
      isLoad={false}
      isLoading={false}
      isNoResultsOption={false}
      isPlaceholder={false}
      isSelected={false}
      key="a"
      keyHandler={[Function]}
      onClick={[Function]}
      sendRef={[Function]}
      setViewMoreNextIndex={[Function]}
      value="fake-regiona"
    >
      fake-regiona
    </SelectOption>
    <SelectOption
      className=""
      component="button"
      index={0}
      inputId=""
      isChecked={false}
      isDisabled={false}
      isFavorite={null}
      isLastOptionBeforeFooter={[Function]}
      isLoad={false}
      isLoading={false}
      isNoResultsOption={false}
      isPlaceholder={false}
      isSelected={false}
      key="b"
      keyHandler={[Function]}
      onClick={[Function]}
      sendRef={[Function]}
      setViewMoreNextIndex={[Function]}
      value="fake-regionb"
    >
      fake-regionb
    </SelectOption>
    <SelectOption
      className=""
      component="button"
      index={0}
      inputId=""
      isChecked={false}
      isDisabled={false}
      isFavorite={null}
      isLastOptionBeforeFooter={[Function]}
      isLoad={false}
      isLoading={false}
      isNoResultsOption={false}
      isPlaceholder={false}
      isSelected={false}
      key="c"
      keyHandler={[Function]}
      onClick={[Function]}
      sendRef={[Function]}
      setViewMoreNextIndex={[Function]}
      value="fake-regionc"
    >
      fake-regionc
    </SelectOption>
    <SelectOption
      className=""
      component="button"
      index={0}
      inputId=""
      isChecked={false}
      isDisabled={false}
      isFavorite={null}
      isLastOptionBeforeFooter={[Function]}
      isLoad={false}
      isLoading={false}
      isNoResultsOption={false}
      isPlaceholder={false}
      isSelected={false}
      key="d"
      keyHandler={[Function]}
      onClick={[Function]}
      sendRef={[Function]}
      setViewMoreNextIndex={[Function]}
      value="fake-regiond"
    >
      fake-regiond
    </SelectOption>
    <SelectOption
      className=""
      component="button"
      index={0}
      inputId=""
      isChecked={false}
      isDisabled={false}
      isFavorite={null}
      isLastOptionBeforeFooter={[Function]}
      isLoad={false}
      isLoading={false}
      isNoResultsOption={false}
      isPlaceholder={false}
      isSelected={false}
      key="e"
      keyHandler={[Function]}
      onClick={[Function]}
      sendRef={[Function]}
      setViewMoreNextIndex={[Function]}
      value="fake-regione"
    >
      fake-regione
    </SelectOption>
    <SelectOption
      className=""
      component="button"
      index={0}
      inputId=""
      isChecked={false}
      isDisabled={false}
      isFavorite={null}
      isLastOptionBeforeFooter={[Function]}
      isLoad={false}
      isLoading={false}
      isNoResultsOption={false}
      isPlaceholder={false}
      isSelected={false}
      key="f"
      keyHandler={[Function]}
      onClick={[Function]}
      sendRef={[Function]}
      setViewMoreNextIndex={[Function]}
      value="fake-regionf"
    >
      fake-regionf
    </SelectOption>
  </Select>
  <FormGroupHelperText />
</FormGroup>
`;
